pkg/syncer/syncertest/fake/discovery_client.go

```diff
@@ -19,28 +19,45 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"kpt.dev/configsync/pkg/util/discovery"
+	discovery "k8s.io/client-go/discovery"
+	discoveryutil "kpt.dev/configsync/pkg/util/discovery"
 )
 
 // discoveryClient implements the subset of the DiscoveryInterface used by the
 // Syncer.
 type discoveryClient struct {
 	resources []*metav1.APIResourceList
+	errors    error
 }
 
 // ServerResources implements discovery.ServerResourcer.
 func (d discoveryClient) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
-	return nil, d.resources, nil
+	return nil, d.resources, d.errors
 }
 
-var _ discovery.ServerResourcer = discoveryClient{}
+var _ discoveryutil.ServerResourcer = discoveryClient{}
+
+func discoveryError(err error, gvks ...schema.GroupVersionKind) error {
+	groups := make(map[schema.GroupVersion]error)
+	for _, gvk := range gvks {
+		gv := gvk.GroupVersion()
+		groups[gv] = err
+	}
+	return &discovery.ErrGroupDiscoveryFailed{Groups: groups}
+}
 
 // NewDiscoveryClient returns a discoveryClient that reports types available
-// to the API Server.
+// to the API Server with no errors, or ErrGroupDiscoveryFailed error with the
+// specified wantedErr
 //
 // Does not report the scope of each GVK as no tests requiring a discoveryClient
 // use scope information.
-func NewDiscoveryClient(gvks ...schema.GroupVersionKind) discovery.ServerResourcer {
+func NewDiscoveryClient(wantedErr error, gvks ...schema.GroupVersionKind) discoveryutil.ServerResourcer {
```

Comment:
nit: Should we make a NewDiscoveryClientWithError constructor so we dont need to use a nil first arg and refactor the old use cases? I don't feel strongly.

Reply:
I was doing that previously, a wrapper over the main constructor, I can revert it so the other callers remains as they were.

Reply:
Seems cleaner to make the error argument optional, yeah.

Reply:
Done